File: android/ValleTPV/app/src/main/java/com/valleapp/valletpv/adaptadoresDatos/AdaptadorTicket.java
package com.valleapp.valletpv.adaptadoresDatos;

import android.annotation.SuppressLint;
import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.ImageButton;
import android.widget.TextView;

import com.valleapp.valletpv.interfaces.IControladorCuenta;
import com.valleapp.valletpv.R;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;

/**
 * Created by xbmc on 8/09/14.
 */
public class AdaptadorTicket extends ArrayAdapter<JSONObject> implements View.OnClickListener {

    private final Context context;
    private final java.util.List<JSONObject> values;
    IControladorCuenta controlador;

    public AdaptadorTicket(Context context, ArrayList<JSONObject> values, IControladorCuenta controlador) {
        super(context, R.layout.item_art, (java.util.List<JSONObject>) values);
        this.context = context;
        this.controlador = controlador;
        this.values = values;
    }

    @SuppressLint("DefaultLocale")
    @Override
    public View getView(int position, View convertView, ViewGroup parent) {

            LayoutInflater inflater = (LayoutInflater) context
                    .getSystemService(Context.LAYOUT_INFLATER_SERVICE);
            @SuppressLint("ViewHolder") View rowView = inflater.inflate(R.layout.item_art, parent, false);

        try {
            TextView can = (TextView) rowView.findViewById(R.id.lblCan);
            TextView nombre = (TextView) rowView.findViewById(R.id.lblNombre);
            TextView p = (TextView) rowView.findViewById(R.id.lblPrecio);
            TextView t = (TextView) rowView.findViewById(R.id.lblTotal);
            can.setText(String.format("%s", values.get(position).getString("Can")));
            nombre.setText(String.format("%s",values.get(position).getString("descripcion_t")));
            p.setText(String.format("%01.2f €", values.get(position).getDouble("Precio")));
            t.setText(String.format("%01.2f €", values.get(position).getDouble("Total")));
            ImageButton rm = (ImageButton)rowView.findViewById(R.id.btn_borrar);
            rm.setTag(values.get(position));
            rm.setOnClickListener(this);
        } catch (JSONException e) {
            e.printStackTrace();
        }


        return rowView;
    }

    @Override
    public void onClick(View view) {
        try {
            JSONObject art = (JSONObject)view.getTag();
            String estado = art.getString("Estado");
            if(estado.equals("N")) controlador.borrarArticulo(art);
            else controlador.clickMostrarBorrar(art);
        } catch (JSONException e) {
            e.printStackTrace();
        }
    }
}

File: android/ValleTPV/ValleTPV/src/main/java/com/valleapp/valletpv/adaptadoresDatos/AdaptadorTicket.kt
package com.valleapp.valletpv.adaptadoresDatos

import android.content.Context
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.ImageButton
import android.widget.TextView
import com.valleapp.valletpv.interfaces.IControladorCuenta
import com.valleapp.valletpv.R
import org.json.JSONException
import org.json.JSONObject
import java.util.Locale

class AdaptadorTicket(
    private val context: Context,
    private val values: ArrayList<JSONObject>,
    private val controlador: IControladorCuenta
) : ArrayAdapter<JSONObject>(context, R.layout.item_art, values), View.OnClickListener {

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        val viewHolder: ViewHolder
        val rowView: View

        if (convertView == null) {
            val inflater = LayoutInflater.from(context)
            rowView = inflater.inflate(R.layout.item_art, parent, false)

            viewHolder = ViewHolder(
                rowView.findViewById(R.id.lblCan),
                rowView.findViewById(R.id.lblNombre),
                rowView.findViewById(R.id.lblPrecio),
                rowView.findViewById(R.id.lblTotal),
                rowView.findViewById(R.id.btn_borrar)
            )

            rowView.tag = viewHolder
        } else {
            rowView = convertView
            viewHolder = rowView.tag as ViewHolder
        }

        val item = values[position]

        try {
            viewHolder.can.text = item.getString("Can")
            viewHolder.nombre.text = item.getString("descripcion_t")
            viewHolder.p.text = String.format(Locale.getDefault(),"%01.2f €", item.getDouble("Precio"))
            viewHolder.t.text = String.format(Locale.getDefault(),"%01.2f €", item.getDouble("Total"))

            viewHolder.rm.tag = item
            viewHolder.rm.setOnClickListener(this)
        } catch (e: JSONException) {
            Log.e("AdaptadorTicket", e.message ?: "Error al procesar JSON")
        }

        return rowView
    }

    override fun onClick(view: View) {
        try {
            val art = view.tag as JSONObject
            val estado = art.getString("Estado")
            if (estado == "N") controlador.borrarArticulo(art)
            else controlador.clickMostrarBorrar(art)
        } catch (e: JSONException) {
            Log.e("AdaptadorTicket", e.message ?: "Error en onClick")
        }
    }

    // ViewHolder para mejorar el rendimiento
    private class ViewHolder(
        val can: TextView,
        val nombre: TextView,
        val p: TextView,
        val t: TextView,
        val rm: ImageButton
    )
}
